core/dbt/tests/adapter.py

```diff
@@ -0,0 +1,193 @@
+        with adapter.connection_named("__test"):
+            for schema in schemas:
+                found_relations.extend(adapter.list_relations_without_caching(schema))
+
+    for key, value in relation_to_type.items():
+        for relation in found_relations:
+            # this might be too broad
+            if relation.identifier == key:
+                assert relation.type == value, (
+                    f"Got an unexpected relation type of {relation.type} "
+                    f"for relation {key}, expected {value}"
+                )
+
+
+def check_relations_equal(adapter, relation_names):
+    if len(relation_names) < 2:
+        raise TestProcessingException(
+            "Not enough relations to compare",
+        )
+    relations = [relation_from_name(adapter, name) for name in relation_names]
+
+    with patch.object(providers, "get_adapter", return_value=adapter):
+        with adapter.connection_named("_test"):
+            basis, compares = relations[0], relations[1:]
+            columns = [c.name for c in adapter.get_columns_in_relation(basis)]
+
+            for relation in compares:
+                sql = adapter.get_rows_different_sql(basis, relation, column_names=columns)
+                _, tbl = adapter.execute(sql, fetch=True)
+                num_rows = len(tbl)
+                assert (
+                    num_rows == 1
+                ), f"Invalid sql query from get_rows_different_sql: incorrect number of rows ({num_rows})"
+                num_cols = len(tbl[0])
+                assert (
+                    num_cols == 2
+                ), f"Invalid sql query from get_rows_different_sql: incorrect number of cols ({num_cols})"
+                row_count_difference = tbl[0][0]
+                assert (
+                    row_count_difference == 0
+                ), f"Got {row_count_difference} difference in row count betwen {basis} and {relation}"
+                rows_mismatched = tbl[0][1]
+                assert (
+                    rows_mismatched == 0
+                ), f"Got {rows_mismatched} different rows between {basis} and {relation}"
+
+
+def get_unique_ids_in_results(results):
+    unique_ids = []
+    for result in results:
+        unique_ids.append(result.node.unique_id)
+    return unique_ids
+
+
+def check_result_nodes_by_name(results, names):
+    result_names = []
+    for result in results:
+        result_names.append(result.node.name)
+    assert set(names) == set(result_names)
+
+
+def check_result_nodes_by_unique_id(results, unique_ids):
+    result_unique_ids = []
+    for result in results:
+        result_unique_ids.append(result.node.unique_id)
+    assert set(unique_ids) == set(result_unique_ids)
+
+
+def update_rows(adapter, update_rows_config):
+    """
+    {
+      "name": "base",
+      "dst_col": "some_date"
+      "clause": {
+          "type": "add_timestamp",
+          "src_col": "some_date",
+       "where" "id > 10"
+    }
```

Comment:
This example is really helpful! I feel like we might want to do this in more places like adding the doc string. For me a lot of these things are new so it was harder to follow the function without an example arg.

Reply:
I added a few more doc strings.